### DISCUSSION TITLE
如何移植Anime4k GLSL shader？

### DISCUSSION THREAD

**Discussion by WHMHammer** [Blinue/Magpie#265] [Dev Channel]
### Question 问题

请问你们是如何移植Anime4k GLSL shader的啊？我想自行添加其它shader。我先参考了这篇[文档](https://docs.microsoft.com/en-us/windows/uwp/gaming/glsl-to-hlsl-reference)进行了类型替换，但是替换后仍无法运行。我又对比了Magpie中的[Anime4K_Restore_L.hlsl](https://github.com/Blinue/Magpie/blob/main/Effects/Anime4K_Restore_L.hlsl)与Anime4k中的[Anime4K_Restore_CNN_L.glsl](https://github.com/bloc97/Anime4K/blob/master/glsl/Restore/Anime4K_Restore_CNN_L.glsl)，仍然没有任何头绪。

**Comment by Blinue** [comment]
Anime4K 基于 [mpv hook](https://mpv.io/manual/stable/#options-glsl-shaders)，Magpie 基于 [MagpieFX](https://github.com/Blinue/Magpie/wiki/%E8%87%AA%E5%AE%9A%E4%B9%89%E6%95%88%E6%9E%9C%EF%BC%88MagpieFX%EF%BC%89)。要做移植工作需要有一定知识储备。

推荐你从最简单的 Anime4K_3D_Upscale 开始
Anime4K -> https://github.com/bloc97/Anime4K/blob/master/glsl/Upscale/Anime4K_3DGraphics_Upscale_x2_US.glsl
Magpie -> https://github.com/Blinue/Magpie/blob/main/Effects/Anime4K_3D_Upscale_US.hlsl

**Comment by WHMHammer** [comment]
谢谢！
